lib/export-api.js

```diff
@@ -99,112 +98,103 @@ module.exports = function (channelName, manifest, methods, globalPermissionCheck
       )
       return
     }
-    if (type == 'readable') {
-      // call readable
-      let stream
+
+    var streamTypes = {
+      readable: createReadableEvents,
+      writable: createWritableEvents,
+      duplex: createDuplexEvents
+    }
+
+    if (streamTypes[type]) {
+      return await handleStream(event, method, requestId, args, streamTypes[type], send)
+    }
+
+    api.emit('error', new Error(`Invalid method type "${type}" for "${methodName}"`), arguments)
+  })
+
+  async function handleStream (event, method, requestId, args, createStreamEvents, send) {
+    // call duplex
+    let stream
+    try {
+      stream = method.apply(event, args)
+      if (!stream) {
+        event.returnValue = { success: false }
+        return
+      }
+    } catch (e) {
+      event.returnValue = { error: e.message }
+      return
+    }
+
+    // handle promises
+    if (stream && stream.then) {
+      event.returnValue = { success: true }
       try {
-        stream = method.apply(event, args)
-        if (!stream) {
-          event.returnValue = { success: false }
-          return
-        }
+        stream = await stream // wait for it
       } catch (e) {
-        event.returnValue = { error: e.message }
+        event.sender.send(channelName, 'stream-error', requestId, '' + e)
```

Comment:
I was going to ask, "Any reason you're not using `send()` here?" But then I realized that's what the existing code did. I think that was a mistake on my part, so let's leave it and I'll do a pass after.

I also think there's a mistake in using the `send()` helper in all cases, because `send()` unregisters the `'did-navigate'` listener, and that should only happen when the stream/transaction is closed. Again, that's me. I'll take a look after this merges.